docs/proposals/api-group.md

```diff
@@ -0,0 +1,154 @@
+
+4. Exposing the API groups and versions supported by the server. This is required to develop a dynamic client.
+
+5. Laying the basis for [API Plugin](../../docs/design/extending-api.md)/
+
+6. Keeping the user interaction easy. For example, we should allow users to omit group name when using kubectl if there is no ambiguity.
+
+
+## Bookkeeping for groups
+
+1. No changes to API spec:
+
+  Currently many internal structures, such as RESTMapper and Scheme, are indexed and retrieved by APIVersion. For a fast implementation targeting the v1.1 deadline, we will concatenate group with version, in the form of "group/version", and use it where a version string is expected, so that many code can be reused. This also means we will not add a new field to the API spec, we will use TypeMeta.APIVersion to hold "group/version".
+
+  For backward compatibility, v1 objects belong to the group with an empty name, so existing v1 config files will remain valid.
+
+2. conversion.Scheme:
+
+  The key of conversion.Scheme.versionMap for versioned types will be "group/version". For now, the internal version types of all groups will be registered to versionMap[""], as we don't have any identically named kinds in different groups yet. In the near future, internal version types will be registered to versionMap["group/"], and pkg/conversion#Scheme.InternalVersion will have type []string.
+
+3. meta.RESTMapper:
+
+  Each group will have its own RESTMapper (of type DefaultRESTMapper), and these mappers will be registered to pkg/api#RESTMapper (of type MultiRESTMapper).
+
+  To support identically named kinds in different groups, We need to expand the input of RESTMapper.VersionAndKindForResource from (resource string) to (group, resource string). If group is not specified and there is ambiguity (i.e., the resource exists in multiple groups), an error should be returned to force the user to specify the group.
+
+## Server-side implementation
+
+1. resource handlers' URL:
+
+  We will force the URL to be in the form of prefix/group/version/...
+
+  Prefix is used to differentiate API paths from other paths like /healthz. Groups really should use the same prefix="apis", except when backward compatibility requires otherwise. Specifically,
+
+    * for /api/v1, we set the prefix="api" (which is populated from cmd/kube-apiserver/app#APIServer.APIPrefix), group="", version="v1", so the URL remains to be /api/v1.
+
+    * for new kube API groups, we will set the prefix="apis" (we will add a field in type APIServer to hold this prefix), group=GROUP_NAME, version=VERSION. For example, the URL of the experimental resources will be /apis/experimental/v1alpha1.
+
+    * for OpenShift v1 API, because it's currently registered at /oapi/v1, to be backward compatible, OpenShift may choose prefix="osapi", group="", or prefix="", group="osapi".
```

Comment:
It's actually oapi in our v1z

On Aug 28, 2015, at 4:42 PM, Brian Grant <notifications@github.com> wrote:

In docs/proposals/api-group.md
<https://github.com/kubernetes/kubernetes/pull/12951#discussion_r38242117>:

> +
> +  To support identically named kinds in different groups, We need to expand the input of RESTMapper.VersionAndKindForResource from (resource string) to (group, resource string). If group is not specified and there is ambiguity (i.e., the resource exists in multiple groups), an error should be returned to force the user to specify the group.
> +
> +## Server-side implementation
> +
> +1. resource handlers' URL:
> +
> +  We will force the URL to be in the form of prefix/group/version/...
> +
> +  Prefix is used to differentiate API paths from other paths like /healthz. Groups really should use the same prefix="apis", except when backward compatibility requires otherwise. Specifically,
> +
> +    * for /api/v1, we set the prefix="api" (which is populated from cmd/kube-apiserver/app#APIServer.APIPrefix), group="", version="v1", so the URL remains to be /api/v1.
> +
> +    * for new kube API groups, we will set the prefix="apis" (we will add a field in type APIServer to hold this prefix), group=GROUP_NAME, version=VERSION. For example, the URL of the experimental resources will be /apis/experimental/v1alpha1.
> +
> +    * for OpenShift v1 API, because it's currently registered at /oapi/v1, to be backward compatible, OpenShift may choose prefix="osapi", group="", or prefix="", group="osapi".

typo: s/oapi/osapi/

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12951/files#r38242117>.